### ISSUE TITLE
Comments from Martin Paljak

### ISSUE THREAD

**Issue by rigow** [w3c/websec#8]
From the list of specifics, I envision two interesting things:
- Secure Hardware IO <- which should cover all our current use cases,
  with the current wording (is _not_ bound to origin)
- Secure Worker <- makes me wonder a bit, what would be the use case and
  what kind of hardware do you envision? TEE protected JS execution? Or
  the current native messaging extended to include a secure environment on
  the other end?

The scope is OK, but I'd like to keep it minimalistic to be able to
actually achieve anything. Right now I see that a lot could be made if a
standardised NativeMessaging solution would be the target, with domain
specific IDL-s and "web intents" style selection of pre-installed
providers (be it your bank, bitcoin vendor, eID vendor, government or
whatnot)

For success criteria, I have an addition:

Engage _in a successful_ dialog with the governmental identity systems
that require a higher level of security

**Comment by wayneca** [comment]
> Secure Worker <- makes me wonder a bit, what would be the use case and 
> what kind of hardware do you envision? TEE protected JS execution? Or the 
> current native messaging extended to include a secure environment on the other end?

There are lots of different hardware secure environments.  These specs likely would abstract those.  They would be used in the underlying implementation.

Likely implementation is the Browser proxying requests to hardware it knows how to deal with.  The communication for Secure Workers would be through message APIs like for Web Workers.  The Browser would know how to deliver that message to the underlying secure hardware implementation. That wouldn't be part of the spec.

**Comment by wayneca** [comment]
> Right now I see that a lot could be made if a
> standardised NativeMessaging solution would be the target, with domain
> specific IDL-s and "web intents" style selection of pre-installed
> providers (be it your bank, bitcoin vendor, eID vendor, government or
> whatnot)

This isn't about web intent style messaging to native code.  There are APIs that the Browser (UA) implements and the Browser knows how to interact with the underlying implementation, so the spec doesn't have to say how that's done.  So, a general mechanism for talking to arbitrary code outside the Browser is not needed (and not in scope).  

Any kind of general passing arbitrary messages to any arbitrary code from random web pages is beyond what's envisioned here.  This is simply give a file containing code that you want executed in a secure environment and a message passing scheme back to the usual code in a Web page (and the more general mechanism for determining if you trust the hardware service provider).

**Comment by cyberphone** [comment]
Before Intel actually provides a specification anything is in scope

**Comment by cyberphone** [comment]
Before Intel actually provides a specification of "something" and the WG is officially launched, virtually "anything" related (including Native Messaging) is by definition in scope

**Comment by cyberphone** [comment]
The browser vendors have anyway decided to support Native Messaging:
https://lists.w3.org/Archives/Public/public-webappsec/2015Oct/0071.html

**Comment by cyberphone** [comment]
Before Intel actually provides a specification of "something" and the WG is officially launched, virtually "anything" related (including Native Messaging) is by definition in scope.

The browser vendors have anyway decided to support Native Messaging:
https://lists.w3.org/Archives/Public/public-webappsec/2015Oct/0071.html

Intel is BTW also a user of Web-2-Native solutions:
http://ipt.intel.com/Home/How-it-works/network-security-identity-management/ipt-with-near-field-communications

**Comment by cyberphone** [comment]
Before Intel actually provides a specification of "something" and the WG is officially launched, virtually "anything" related (including Native Messaging) is by definition in scope.

The browser vendors have anyway decided to support Native Messaging:
https://lists.w3.org/Archives/Public/public-webappsec/2015Oct/0071.html

Intel is BTW also a user of Web-2-Native solutions:
http://ipt.intel.com/Home/How-it-works/network-security-identity-management/ipt-with-near-field-communications

There is no magic going on here:
https://cyberphone.github.io/openkeystore/resources/docs/web2native-bridge.pdf#page=4

**Comment by wayneca** [comment]
@cyberphone 

> Before Intel actually provides a specification of "something" 
> and the WG is officially launched, virtually "anything" related 
> (including Native Messaging) is by definition in scope.

I thought you misunderstood the description of secure worker and thought it would require what you call "native messaging" and that you thought it would need to be added to the Charter.  I was trying to explain that the proposed APIs do not require that so it isn't required by the Charter as written now.  

I assume now that what you meant is what has been brought up in multiple existing working groups, the idea to drop various W3C specs or groups and to instead create a way for Web pages to start and communicate with arbitrary native applications already installed on devices.  So, don't do Web NFC and instead have "native messaging" and talk to a native app that can do it.  Don't do web payments, just ... There was a Community Group started for that [https://www.w3.org/community/extended-web/](https://www.w3.org/community/extended-web/).  That CG currently 2 members and no activity ever on the mailing list.

As I understand it now, what you mean is to drop the current items and instead specify "native messaging" in this proposed WG.  If W3C were to go that way, I'd think it would be in the Web Platform WG, not this new proposed group that is about using security services in specialized security hardware.  "Native messaging" is not about using special hardware security support.  So it is not what is being talked about here.

Web Intents (which was in Web Apps which is now moved into Web Platform) was about being able to register services (usually provided by a Web page in another tab) that could be registered and then used by other Web pages.  An additional part of that would have been to also allow native apps to register in some way to similarly provide services locally, rather from a server.  That would require Browsers to provide some way for native apps to register and some way to pass messages to native applications (likely different on each OS because of restrictions on interprocess communication in some systems).  That whole effort didn't work out.  It's not abandoned and it likely will be back in some different form.  This proposed hardware security group doesn't seem the right forum to me for addressing that since the idea isn't that it would need some kind of security hardware.

**Comment by wayneca** [comment]
I should have mentioned:  The current version of the charter includes permitting RESTful Web Services versions of the APIs.  So there would be a message based format for these APIs.  If, in the future in some other W3C WG, Web Web Intents with service registration by trusted native apps was defined, then the message formats defined in this group could be used.  Nothing different would need to be done in this proposed Working Group at all, so no need to change the Charter.  But, that means some more appropriate Working Group defining Web Intents that allow access to services from native applications.

**Comment by cyberphone** [comment]
If we stick to the kind of applications that most people associate with security hardware (authentication and payments), I remain unconvinced that a "pure" Web effort will be able providing something which matches the functionality offered by (the nowadays deprecated) browser plugins.

In fact, the browser vendors are about to throw out existing functionality like on-line key-generation (https://lists.w3.org/Archives/Public/www-tag/2015Sep/0000.html), with the ultimate goal dropping support for HTTPS client-certificate authentication which is typically used for e-government service access.

Regarding Web NFC (https://w3c.github.io/web-nfc/), the current spec essentially outlines a complex replacement of QR codes.  Payments are explicitly out of scope and authentication isn't even mentioned.  The Web NFC security model doesn't match these applications either.

In summary: W3C runs a bunch of highly disparate and piecemeal efforts, making FIDO alliance a more logical forum for hardware security standardization.  Although Native Messaging does not require security hardware, it is a very powerful way of adding such capabilities to the Web.  That is, hardware security standards can be introduced (and in a timely fashion) on top of Native Messaging.

**Comment by cyberphone** [comment]
It is important keeping in mind that Google is the to-date only entity that managed creating a hardware security solution (U2F) that is integrated the whole way up to the Web.  Then then got close to the entire industry backing it as well.

To succeed with such an undertaking your need:
* A strong vision ("what problem you want to solve")
* Management buy-in
* A reasonably well tested technical solution
* Browser-vendor support

Non of these qualities appear to be applicable to the current charter.

http://poulpita.com/2015/02/11/fido-alliance-an-amazing-standardization-story/

**Comment by cyberphone** [comment]
It is important keeping in mind that Google is the to-date only entity that managed creating a hardware security solution (U2F) that is integrated the whole way up to the Web. Then then got close to the entire industry backing it as well.  http://poulpita.com/2015/02/11/fido-alliance-an-amazing-standardization-story/

To succeed with such an undertaking your need:

* A strong vision ("What problem you want to solve")
* Management buy-in
* A reasonably well-tested technical solution
* Browser-vendor support

None of these qualities appear to be applicable to the current charter.

**Comment by cyberphone** [comment]
It is important keeping in mind that Google is the (to-date) only entity that have managed creating a hardware security solution (U2F) that is integrated the whole way up to the Web.  Then they also got close to the entire industry backing it.
http://poulpita.com/2015/02/11/fido-alliance-an-amazing-standardization-story/

To succeed with such an undertaking you need:
* A strong vision ("What problem do you want to solve")
* Management buy-in
* A reasonably well-tested technical solution
* Browser-vendor support

None of these qualities seem to be applicable to the current charter.

In all fairness it is worth mentioning another factor contributing to the success of U2F: Google have no business interests whatsoever in the actual solution, only in its use.  The "real" Hardware Security folks OTOH, are riddled by business- and IPR-considerations.

**Comment by cyberphone** [comment]
It is important keeping in mind that Google is the (to-date) only entity that have managed creating a hardware security solution (U2F) that is integrated the whole way up to the Web. Then they also got close to the entire industry backing it.
http://poulpita.com/2015/02/11/fido-alliance-an-amazing-standardization-story/

To succeed with such an undertaking you need:

* A strong vision ("What problem do you want to solve")
* Management buy-in
* A reasonably well-tested technical solution
* Browser-vendor support

None of these qualities seem to be applicable to the current charter.

In all fairness it is worth mentioning another factor contributing to the success of U2F: Google have no business interests whatsoever in the actual solution, only in its use. The "real" Hardware Security folks OTOH, are riddled by business- and IPR-considerations making standardization very difficult.  Native Messaging respects this situation :-)

**Comment by cyberphone** [comment]
It is important keeping in mind that Google is the (to-date) only entity that have managed creating a hardware security solution (U2F) that is integrated the whole way up to the Web. Then they also got close to the entire industry backing it.
http://poulpita.com/2015/02/11/fido-alliance-an-amazing-standardization-story/

To succeed with such an undertaking you need:
- A strong vision ("What problem do you want to solve")
- Management buy-in
- A reasonably well-tested technical solution
- Browser-vendor support

_None of these qualities seem to be applicable to the current charter_.

In all fairness it is worth mentioning another factor contributing to the success of U2F: Google have no business interests whatsoever in the actual solution, only in its use. The "real" Hardware Security folks OTOH, are riddled with business- and IPR-considerations making standardization very difficult. Native Messaging respects this situation :-)

**Comment by cyberphone** [comment]
Google is now about creating yet another proprietary Native Messaging solution
https://github.com/w3c/webpayments/issues/42#issuecomment-166705416
that only works for payments.

**Comment by vgalindo** [comment]
@rigow I suggest we close this issue.

**Comment by vgalindo** [comment]
@rigow lets create an additional sentence to mention we will analyse mechanisms such as native messaging.

**Comment by wayneca** [comment]
@vgalindo I'm not sure what that would mean.  It shouldn't be in the restricted scope section.  "native messaging" has nothing at all to do with secure hardware.  It's a more general capability of somehow finding and sending messages from Web pages to native applications on the user's device.  It is often suggested as an alternative to standardization, e.g. just directly use a popular native application.  Web Intents was one way to do that.  That is still under discussion elsewhere.  An example of this is using localhost and a local web server to provide a way to communicate with an installed application.  Web App Sec WG or Web Platforms WG, but what would this WG analyse related to that?  (it would defeat the purpose to only have it available if you had some specific security hardware).

**Comment by cyberphone** [comment]
@wcarr Web Intents and localhost schemes are unsuitable for security applications since they do not provide a security context to the invoked application.

The purpose of at least my take on Native Messaging is not to eliminate the need for Web standards, but to enable a more flexible way of creating such.  That is, there would be one master standard, and any number of tightly scoped slave-standards building on that.

If OTOH, somebody can come up with a implementable and reasonably universal scheme for interacting with security hardware from Web applications, nobody would be happier than me.  I have actually tried and it didn't work, it pretty soon got way too complex unlike my Native Messaging stuff which is quite simple.

**Comment by wayneca** [comment]
@cyberphone This proposed WG is not trying to provide a scheme for directly interacting with arbitrary services in any secure hardware.  It's meant to work on a specific set of APIs that the Browser will implement and the Browser will be the one to connect to secure hardware it knows how to connect to.  The API itself doesn't have to include how that underlying Browser implementation works.

**Comment by cyberphone** [comment]
@wcarr I'm aware of that.  However, there's no proof that this is actually technically feasible unless we are talking about rather specific schemes limited by SOP like U2F.  If the latter is not the case, it in reality means that the WG in progress have found a solution to the following "enigma":
https://cyberphone.github.io/openkeystore/resources/docs/web2native-bridge.pdf#page=3

Unless you are holding back on some important piece of information, I think you when getting closer to the target will end-up with complete services with UI and all which most likely will stall WG progress indefinitely since it departs from the current quest for actually _removing_ such from the browser, with HTML5's **keygen** as first example: https://lists.w3.org/Archives/Public/www-tag/2015Sep/0000.html

Personally (after spending 6 years in TrustedComputingGroup who failed providing the consumer market with something tangible), I'm rather trying to establish useful stuff in this space _outside_ of standardization, and _if successful_, leave it to SDO folks to cater for.  JavaScript signatures is one such effort which clearly competes (quite favorable) with its IETF counterpart:
https://cyberphone.github.io/openkeystore/resources/docs/jcs.html#ECMAScript_Compatibility_Mode

**Comment by wayneca** [comment]
@cyberphone 
[https://cyberphone.github.io/openkeystore/resources/docs/web2native-bridge.pdf#page=3](https://cyberphone.github.io/openkeystore/resources/docs/web2native-bridge.pdf#page=3)

I think that page does get at the disconnect we're having.  It says:

> A web-application wants to connect to a secure element. Since a browser does not “understand” APDU it can only offer a primitive security prompt.

and

> Since the browser cannot know what the application is about to do with signed data, it can only offer a primitive security prompt.

Nothing like that is proposed in the restricted scope sections.  A web page would NOT form any arbitrary connection to a secure element.  It isn't an API to allow Web pages to use APDUs directly.  It's very specific APIs like those that already exist without this to do some specific things.  Like the crypto APIs or storage APIs.  There is no proposal to make arbitrary connections to secure hardware services or to send that hardware arbitrary messages.  This isn't like the Secure Element approach that was in SysApps.

In a concrete example, there could be an API to encrypt.  That doesn't say anything about how the Browser connects to secure hardware to carry it out and it doesn't provide any direct connection to any specific hardware.  It doesn't say anything about how the Browser implements the API or how it connects with secure hardware.  That is part of the Browser implementation.

If there was "native messaging" and some standard native messaging protocol for some service, the Browser's implementation could use that.  But, it would be irrelevant to the Web page so not part of these specs. 

I think more may be being read into the restricted scope items than are actually there.  This is basically an API for use in a Web page to let the server behind that page determine that it trusts some specific secure hardware service (like executing JavaScript code in a secure context and letting it send messages back to Javascript on the web page).  The Browser mediates the interaction and it knows how to communicate with the secure hardware.  The Web page doesn't need to know anything about that at all.

So, the Web page code doesn't need to be able to connect to arbitrary native code on the device (and the Browser already can do that).

There is some work on safely using localhost to connect to a server on a device from a Web page.  That's interesting.   And since providing Web services versions of the APIs is part of this spec, if done those could be used by that.  But, it doesn't mean that "native messaging" itself should be defined by a hardware security WG (when the general concept doesn't at all depend on any special hardware).  WebAppSec seems a fine place for that.

**Comment by martinpaljak** [comment]
How is mentioning native messaging not OK but mentioning web services is OK? Both are technologies for achieving a "transport layer".

There are two approaches: realistic iterative improvement and architecture astronautics. There can't be a "web-way" that contradicts everything happening in real life, nor should we be creating any (that would be the architecture astronautics)

For example, lets take payments. Is the goal to facilitate commerce on the web by providing payment frameworks and networks (my bank, Visa, Bitcoin, whatnot) means for building their payment solutions, or is it to define a new payment type "web-payment" to be implemented by interested parties (my bank, shops, browsers etc) ? I would like to think that NO. While it would be "open" ("everybody can implement it, if they want, so it is open and democratic") it would be that in a false sense and actually giving a "my way or the highway" solution. Should the web get involved and define in global finance? Is it something that should be for a small task force to work on and more precisely, achieve meaningful goals in <7 years? I'd guess no. I don't want the browser to define what a payment is and how it should work.

So instead of providing purpose-built API-s ("web-payments" API instead of "methods for implementing payments on the web, that can include hardware") and trying to overload "secure", "trust", "identity" in a "webish way", it makes more sense to provide a framework for implementations and let the interest groups build their API-s and frameworks on top of it. To compare it to the internet: we need a solid IP stack, not a plethora of protocols on top of it. The question is where to put the development effort: into the browser (would be not moving with the pace of innovation in either identity, payments or eg signature fields and probably be biased) or elsewhere.

**Comment by cyberphone** [comment]
Native messaging is happening on a grand scale like the Chrome-to-Android interface suggested (and recently also implemented) by Google.

Google is effectively leading this development through their since long shipping native messaging solution in Chrome desktop which Mozilla and Microsoft have publicly indicated that they intend to support.

Since none of the existing WGs or TAG have shown interest in this kind of Web technology; it will rather become a de-facto standard.

As @martinpaljak points out, the ability to extend the Web through other means than time-consuming and fairly non-democratic standardization processes is an important feature that eventually could make the (for the "market") uninteresting war between the Web and Native a thing of the past.

**Comment by wayneca** [comment]
@cyberphone Just to be clear [Google's Native Messaging](https://developer.chrome.com/extensions/nativeMessaging) is an API for use only by Browser extensions or separate trusted applications.  Has Google proposed that API for use by arbitrary Web pages?

@martinpaljak 

> How is mentioning native messaging not OK but mentioning web services is OK? Both are technologies for achieving a "transport layer"

@martinpaljak  RESTful Web Services are something that already exists that Web pages can already use through W3C specs.  Native messaging is something that would need to be defined somewhere.  It's a general capability that previously was considered as part of WebApps Web Intents.  It is a very general capability that should be defined in one of the already existing WGs like the Web Platform WG or WebAppSec WG.  It doesn't seem the type of thing to be created by this, fairly special purpose WG.

That said, W3C Web Application Security WG have been discussing a safe way that would enable public Web pages to use services from a local Web Server as well as access to private networks.  [https://mikewest.github.io/cors-rfc1918/](https://mikewest.github.io/cors-rfc1918/), [https://lists.w3.org/Archives/Public/public-webappsec/2016Jan/0000.html](https://lists.w3.org/Archives/Public/public-webappsec/2016Jan/0000.html)

That is better in the Web Application Security WG, than to be developed in this proposed WG.  If it already existed, the Web Services interface that is already in the charter could be used with a local server on the device - so nothing special would need to be done.

There's also a Web Payments WG.  This proposal isn't for that either.

**Comment by cyberphone** [comment]
@wcarr First off, these are _very difficult topics_ which haven't been discussed much in public forums.

> @cyberphone Just to be clear Google's Native Messaging is an API for use only by Browser extensions or separate trusted applications. Has Google proposed that API for use by arbitrary Web pages?

Anybody can write a browser extension and getting it published on the Chrome store.  The called native application is not a part of the publishing (or granting) process.  Extensions may be called from _any_ Web page. That is, there's no concept of "trusted application" in the current scheme  which is one of the reasons I designed a new scheme where _trust_, _security_, _user interface_, and _deployment_ issues were considered from the beginning.

AFAICT, accessing RESTFul Web Services running on "localhost" suffer from equally serious security and usability issues. _It is rather a less powerful version of native messaging_.  I don't see that this concept has any direct links to security hardware either.

IMO, the core problem here is a lack of described (and analyzed) use cases.  That somebody wants to "encrypt data" isn't enough. The applications we (presumably) are targeting typically have an UI.  Where does the UI live? Well, there are virtually tons of unanswered questions out there that at least I wouldn't like to deal with before committing to any kind of charter or API.  In my book this is a _research project_.  My take on native messaging has though (IMO) passed that stage of so I'm looking for possibilities to implement this in Android.

Security analysis of Google's take on native messaging:
https://lists.w3.org/Archives/Public/public-webappsec/2015Oct/0071.html

**Comment by rigow** [comment]
@vgalindo and me have talked with @martinpaljak on the phone to resolve the issue. I think it is understood that we want to create a layer of interoperability that, in its implementation, requires some local code, but is not just a pipeline into execution of local code. So I think this can be closed now
